sdk/src/instruction_processor_utils.rs

```diff
@@ -51,17 +54,70 @@ macro_rules! solana_entrypoint(
 ///   Ok(())
 /// }
 ///
-/// solana_sdk::declare_program!(
-///     "My!!!11111111111111111111111111111111111111",
+/// declare_program!(
+///     "My11111111111111111111111111111111111111111",
 ///     solana_my_program,
 ///     my_process_instruction
 /// );
+/// # }
+/// ```
+/// ```
+/// # // wrapper is used so that the macro invocation occurs in the item position
+/// # // rather than in the statement position which isn't allowed.
+/// # mod item_wrapper {
+/// use solana_sdk::account::KeyedAccount;
+/// use solana_sdk::instruction::InstructionError;
+/// use solana_sdk::pubkey::Pubkey;
+/// use solana_sdk::declare_program;
+///
+/// fn my_process_instruction(
+///     program_id: &Pubkey,
+///     keyed_accounts: &mut [KeyedAccount],
+///     data: &[u8],
+/// ) -> Result<(), InstructionError> {
+///   // Process an instruction
+///   Ok(())
+/// }
+///
+/// declare_program!(
+///     solana_sdk::system_program::ID,
+///     solana_my_program,
+///     my_process_instruction
+/// );
+/// # }
 /// ```
 #[macro_export]
 macro_rules! declare_program(
+    ($program_id:ident, $name:ident, $entrypoint:expr) => (
+        declare_program!(@expand $program_id, $name, $entrypoint);
+    );
+    ($program_id:path, $name:ident, $entrypoint:expr) => (
+        declare_program!(@expand $program_id, $name, $entrypoint);
+    );
     ($bs58_string:expr, $name:ident, $entrypoint:expr) => (
         $crate::declare_id!($bs58_string);
+        declare_program!(@declare $name, $entrypoint);
+    );
+    (@expand $program_id:expr, $name:ident, $entrypoint:expr) => (
+        pub static ID: ::solana_sdk::pubkey::Pubkey = $program_id;
```

Comment:
Why not usee `$crate` here for `::solana_sdk'?

Reply:
Ah thanks, my mistake. I missed updating that one after copying from the new proc macro

`$crate` is used in declarative macros and absolute paths are used in procedural macros